### PR TITLE
Refactor Generator Creation and Modification with generic design

### PR DESCRIPTION
**Pull request by ayolab**
Signed-off-by: Ayoub LABIDI <ayoub.labidi@protonmail.com>

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                .withKey("generatorModification")
+                .withDefaultMessage(exc.getMessage())
+                .withValue("id", modificationInfos.getEquipmentId())
+                .withSeverity(TypedValue.ERROR_SEVERITY)
+                .build());
+        }
+    }
+
+    private void modifyGenerator(Generator generator, GeneratorModificationInfos modificationInfos, Reporter subReporter) {
+        subReporter.report(Report.builder()
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java:50]: Move the log after modification

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorCreation.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    @Override
+    public void apply(Network network, Reporter subReporter) {
+        if (modificationInfos == null) {
+            throw new NetworkModificationException(CREATE_GENERATOR_ERROR, "Missing required attributes to create the generator");
+        }
+        // create the generator in the network
+        VoltageLevel voltageLevel = ModificationUtils.getInstance().getVoltageLevel(network, modificationInfos.getVoltageLevelId());
+        if (voltageLevel.getTopologyKind() == TopologyKind.NODE_BREAKER) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/GeneratorCreation.java:51]: Add a method `createGeneratorInNodeBreaker`

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorCreation.java (hunk 2 above)
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/modifications/GeneratorCreation.java:51]: Done

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java (hunk 1 above)
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java]: Done

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void apply(Network network, Reporter subReporter) {
+        if (modificationInfos == null) {
+            throw new NetworkModificationException(MODIFY_GENERATOR_ERROR, "Missing required attributes to modify the equipment");
+        }
+        try {
+            Generator generator = ModificationUtils.getInstance().getGenerator(network, modificationInfos.getEquipmentId());
+            // modify the generator in the network
+            modifyGenerator(generator, modificationInfos, subReporter);
+        } catch (NetworkModificationException exc) {
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java]: Done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/BasicEquipmentModificationEntity.java [hunk 4]
```diff
@@ -31,9 +33,26 @@ public class BasicEquipmentModificationEntity extends EquipmentModificationEntit
...
+        assignAttributes(modificationInfos);
+    }
+
+    @Override
+    public void update(ModificationInfos modificationInfos) {
+        super.update(modificationInfos);
+        assignAttributes((BasicEquipmentModificationInfos) modificationInfos);
+    }
+
+    protected void assignAttributes(BasicEquipmentModificationInfos modificationInfos) {
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/BasicEquipmentModificationEntity.java]: Done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/GeneratorCreationEntity.java [hunk 5]
```diff
@@ -101,37 +104,42 @@ public class GeneratorCreationEntity extends InjectionCreationEntity {
...
+        assignAttributes(generatorCreationInfos);
+    }
+
+    @Override
+    public void update(@NonNull ModificationInfos modificationInfos) {
+        super.update(modificationInfos);
+        assignAttributes((GeneratorCreationInfos) modificationInfos);
+    }
+
+    void assignAttributes(GeneratorCreationInfos generatorCreationInfos) {
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/GeneratorCreationEntity.java]: Done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/InjectionModificationEntity.java [hunk 6]
```diff
@@ -48,4 +51,22 @@ protected InjectionModificationEntity(ModificationType modificationType, String
...
+        assignAttributes(modificationInfos);
+    }
+
+    @Override
+    public void update(ModificationInfos modificationInfos) {
+        super.update(modificationInfos);
+        assignAttributes((InjectionModificationInfos) modificationInfos);
+    }
+
+    protected void assignAttributes(InjectionModificationInfos modificationInfos) {
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/InjectionModificationEntity.java]: Done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/modification/GeneratorModificationEntity.java [hunk 7]
```diff
@@ -93,26 +96,26 @@ public class GeneratorModificationEntity extends InjectionModificationEntity {
...
+        super.update(modificationInfos);
+        assignAttributes((GeneratorModificationInfos) modificationInfos);
     }
 
-    public GeneratorModificationEntity(GeneratorModificationInfos infos) {
-        this(infos.getEquipmentId(), infos.getEquipmentName(), infos.getVoltageLevelId(), infos.getBusOrBusbarSectionId(),
-            infos.getEnergySource(),
-            infos.getMinActivePower(), infos.getMaxActivePower(), infos.getRatedNominalPower(),
-            infos.getActivePowerSetpoint(), infos.getReactivePowerSetpoint(), infos.getVoltageRegulationOn(), infos.getVoltageSetpoint());
+    void assignAttributes(GeneratorModificationInfos generatorModificationInfos) {
```
[the remark this answers is not in the document]
Comment (ayolab) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/modification/GeneratorModificationEntity.java]: Done

File: src/main/java/org/gridsuite/modification/server/modifications/ModificationUtils.java [hunk 8]
```diff
@@ -20,6 +20,8 @@
 import java.util.List;
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/modification/server/modifications/ModificationUtils.java:20]: missing copyright header here

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        return new MatcherGeneratorModificationInfos(generatorModificationInfos);
+    }
+
+    @Override
+    public boolean matchesSafely(GeneratorModificationInfos m) {
+        return super.matchesSafely(m)
+                && m.getEquipmentId().equals(reference.getEquipmentId())
+                && m.getSubstationIds().equals(reference.getSubstationIds())
+                && m.getEquipmentName().equals(reference.getEquipmentName())
+                && Objects.equals(m.getEnergySource(), reference.getEnergySource())
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java:35]: Missing (like in MatcherLoadCreationInfos):
&& m.getVoltageLevelId().equals(reference.getVoltageLevelId())
&& m.getBusOrBusbarSectionId().equals(reference.getBusOrBusbarSectionId())

from InjectionCreationInfos
Maybe we could have MatcherInjectionCreationInfos

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                && Objects.equals(m.getDroop(), reference.getDroop())
+                && Objects.equals(m.getParticipate(), reference.getParticipate())
+                && Objects.equals(m.getRegulatingTerminalId(), reference.getRegulatingTerminalId())
+                && Objects.equals(m.getRegulatingTerminalType(), reference.getRegulatingTerminalType())
+                && Objects.equals(m.getRegulatingTerminalVlId(), reference.getRegulatingTerminalVlId())
+                && Objects.equals(m.getMaximumReactivePower(), reference.getMaximumReactivePower())
+                && Objects.equals(m.getMinimumReactivePower(), reference.getMinimumReactivePower())
+                && Objects.equals(m.getMarginalCost(), reference.getMarginalCost())
+                && Objects.equals(m.getStepUpTransformerReactance(), reference.getStepUpTransformerReactance())
+                && Objects.equals(m.getTransientReactance(), reference.getTransientReactance())
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java:52]: missing check for:
qPercent
voltageRegulationType

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Override
+    protected MatcherGeneratorCreationInfos createMatcher(ModificationInfos modificationInfos) {
+        return MatcherGeneratorCreationInfos.createMatcherGeneratorCreationInfos((GeneratorCreationInfos) modificationInfos);
+    }
+
+    @Override
+    protected void assertNetworkAfterCreation() {
+        assertNotNull(getNetwork().getGenerator("idGenerator2"));
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java:117]: you can also assert the VL update:
assertEquals(1, getNetwork().getVoltageLevel("v1").getGeneratorStream().filter(transformer -> transformer.getId().equals("idGenerator2")).count());

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Override
+    protected void assertNetworkAfterDeletion() {
+        assertNull(getNetwork().getGenerator("idGenerator2"));
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java:122]: same
assertEquals(0, getNetwork().getVoltageLevel("v1").getGeneratorStream().filter(transformer -> transformer.getId().equals("idGenerator2")).count());

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @Override
+    protected MatcherGeneratorCreationInfos createMatcher(ModificationInfos modificationInfos) {
+        return MatcherGeneratorCreationInfos.createMatcherGeneratorCreationInfos((GeneratorCreationInfos) modificationInfos);
+    }
+
+    @Override
+    protected void assertNetworkAfterCreation() {
+        assertNotNull(getNetwork().getGenerator("idGenerator1"));
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java:124]: you can assert the VL update:
assertEquals(1, getNetwork().getVoltageLevel("v2").getGeneratorStream().filter(transformer -> transformer.getId().equals("idGenerator1")).count());

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @Override
+    protected void assertNetworkAfterDeletion() {
+        assertNull(getNetwork().getGenerator("idGenerator1"));
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java:129]: assertEquals(0, getNetwork().getVoltageLevel("v2").getGeneratorStream().filter(transformer -> transformer.getId().equals("idGenerator1")).count());

**Review by dbraquart — CHANGES_REQUESTED** [review]
tests: ok
review: a few minor changes to address

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java (hunk 14 above)
Comment (ayolab) [reply src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java:131]: Done

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java (hunk 13 above)
Comment (ayolab) [reply src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInNodeBreakerTest.java:124]: Done

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java (hunk 12 above)
Comment (ayolab) [reply src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java:124]: Done

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java (hunk 11 above)
Comment (ayolab) [reply src/test/java/org/gridsuite/modification/server/modifications/GeneratorCreationInBusBreakerTest.java:117]: Done

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java (hunk 10 above)
Comment (ayolab) [reply src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java:54]: Done

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java (hunk 9 above)
Comment (ayolab) [reply src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorModificationInfos.java:37]: Done

**Review by dbraquart — APPROVED** [review]

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**